src/dbnode/encoding/proto/iterator.go

```diff
@@ -0,0 +1,878 @@
+	}
+
+	buf := make([]byte, 0, bytesLen)
+	for j := 0; j < int(bytesLen); j++ {
+		b, err := it.stream.ReadByte()
+		if err != nil {
+			return fmt.Errorf(
+				"%s error trying to read byte in readBytes: %v",
+				itErrPrefix, err)
+		}
+		buf = append(buf, b)
+	}
+
+	// TODO(rartoul): Could make this more efficient with unsafe string conversion or by pre-processing
+	// schemas to only have bytes since its all the same over the wire.
+	// https://github.com/m3db/m3/issues/1471
+	schemaFieldType := it.schema.FindFieldByNumber(int32(customField.fieldNum)).GetType()
+	if schemaFieldType == dpb.FieldDescriptorProto_TYPE_STRING {
+		it.lastIterated.TrySetFieldByNumber(customField.fieldNum, string(buf))
+	} else {
+		it.lastIterated.TrySetFieldByNumber(customField.fieldNum, buf)
+	}
+	if err != nil {
+		return fmt.Errorf(
+			"%s error trying to set field number: %d, err: %v",
+			itErrPrefix, customField.fieldNum, err)
+	}
+
+	it.addToBytesDict(i, buf)
+	return nil
+}
+
+// updateLastIteratedWithCustomValues updates lastIterated with the current
+// value of the custom field in it.customFields at index i. This ensures that
+// when we return it.lastIterated in the call to Current() that all the
+// most recent values are present.
+func (it *iterator) updateLastIteratedWithCustomValues(i int) error {
+	if it.lastIterated == nil {
+		it.lastIterated = dynamic.NewMessage(it.schema)
+	}
+
+	var (
+		fieldNum  = it.customFields[i].fieldNum
+		fieldType = it.customFields[i].fieldType
+	)
+
+	switch {
+	case isCustomFloatEncodedField(fieldType):
+		var (
+			val = math.Float64frombits(it.customFields[i].prevFloatBits)
+			err error
+		)
+		if fieldType == cFloat64 {
+			err = it.lastIterated.TrySetFieldByNumber(fieldNum, val)
+		} else {
+			err = it.lastIterated.TrySetFieldByNumber(fieldNum, float32(val))
+		}
+		return err
+
+	case isCustomIntEncodedField(fieldType):
+		switch fieldType {
+		case cSignedInt64:
+			val := int64(it.customFields[i].prevFloatBits)
+			return it.lastIterated.TrySetFieldByNumber(fieldNum, val)
+
+		case cUnsignedInt64:
+			val := it.customFields[i].prevFloatBits
+			return it.lastIterated.TrySetFieldByNumber(fieldNum, val)
+
+		case cSignedInt32:
+			val := int32(it.customFields[i].prevFloatBits)
+			return it.lastIterated.TrySetFieldByNumber(fieldNum, val)
+
+		case cUnsignedInt32:
+			val := uint32(it.customFields[i].prevFloatBits)
+			return it.lastIterated.TrySetFieldByNumber(fieldNum, val)
+
+		default:
+			return fmt.Errorf(
+				"%s expected custom int encoded field but field type was: %v",
+				itErrPrefix, fieldType)
+		}
+	default:
+		return fmt.Errorf(
+			"%s unhandled fieldType: %v", itErrPrefix, fieldType)
+	}
+}
+
+func (it *iterator) readFloatXOR(i int) (floatBits, xor uint64, err error) {
+	xor, err = it.readXOR(i)
+	if err != nil {
+		return 0, 0, err
+	}
+	prevFloatBits := it.customFields[i].prevFloatBits
+	return prevFloatBits ^ xor, xor, nil
+}
+
+func (it *iterator) readXOR(i int) (uint64, error) {
```

Comment:
We should probably refactor the horizontal field encoding logic into some lower level encoding packages so you can re-use certain type encoding logic (in addition to the delta of delta timestamp stuff) across the m3tsz and proto encoders and iterators. 

I guess we can do that when we remove the m3tsz iterator out of here.

Reply:
Yeah right now I think it makes sense for proto to depend on m3tsz since thats pretty much literally what it is logically. Can refactor if that turns out to be the wrong decision